### PR TITLE
Add SIGTERM handling to geographic gatherer

### PR DESCRIPTION
**Pull request by pnuu**
This PR adds SIGTERM handling to geographic gatherer.

After receiving the signal, the gatherer will wait until all active collections have been finished and then exit the program.

### CODE REVIEW

**Comment by pnuu** [comment]
Locally I have an issue with the tests never exiting. The tests pass, but then hang until CTRL+C:
```
======================================================== 14 passed, 22 warnings in 2.22s ========================================================
^CException ignored in: <module 'threading' from '/home/lahtinep/mambaforge/envs/py312/lib/python3.12/threading.py'>
Traceback (most recent call last):
  File "/home/lahtinep/mambaforge/envs/py312/lib/python3.12/threading.py", line 1624, in _shutdown
    lock.acquire()
KeyboardInterrupt: 
/home/lahtinep/mambaforge/envs/py312/lib/python3.12/site-packages/_pytest/pathlib.py:95: PytestWarning: (rm_rf) error removing /tmp/pytest-of-lahtinep/garbage-3a9a0511-3423-47e3-959b-19062ba25739
<class 'OSError'>: [Errno 39] Directory not empty: '/tmp/pytest-of-lahtinep/garbage-3a9a0511-3423-47e3-959b-19062ba25739'
  warnings.warn(
```

The actual functionality seems to work fine when running the main script. At least when there are no collecting going on, which I don't have an easy way to test.

**Comment by pnuu** [comment]
Running the actual script in terminal and killing the process:

```
$ geographic_gatherer.py -p 40000 -n false -i 127.0.0.1:40001 /tmp/geographic_gatherer.ini 
/home/lahtinep/mambaforge/envs/py312/bin/geographic_gatherer.py:4: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  __import__('pkg_resources').require('pytroll-collectors==0.16.0+0.gacc2631.dirty')
Setting timezone to UTC
[INFO: 2024-11-12 08:49:33 : posttroll.backends.zmq.publisher] Publisher for tcp://*:40000 started on port 40000.
[INFO: 2024-11-12 08:49:33 : posttroll.backends.zmq.subscriber] Subscriber adding address tcp://127.0.0.1:40001 with topics ['pytroll://ears/mersi/fy-3d/segment']
[INFO: 2024-11-12 08:49:47 : pytroll_collectors.geographic_gatherer] Caught SIGTERM, shutting down when all collections are finished.
[INFO: 2024-11-12 08:49:47 : pytroll_collectors.geographic_gatherer] Shutting down...
[INFO: 2024-11-12 08:49:47 : pytroll_collectors.geographic_gatherer] Ending the gathering of granules...
[INFO: 2024-11-12 08:49:47 : geographic_gatherer] GeographicGatherer has stopped.
```

File: pytroll_collectors/geographic_gatherer.py [hunk 1]
```diff
@@ -103,8 +106,10 @@ def _setup_triggers(self):
 
     def run(self):
         """Run granule triggers."""
+        signal.signal(signal.SIGTERM, self._handle_sigterm)
         try:
             while True:
+                self._check_sigterm()
```
Comment (mraspaud) [inline pytroll_collectors/geographic_gatherer.py:112]: why not make that function return a boolean and use it in the while statement?

Comment (pnuu) [reply pytroll_collectors/geographic_gatherer.py]: Adjusted in https://github.com/pytroll/pytroll-collectors/pull/157/commits/43b8cc73039b87da43c6252dd83bdf5c861a9b97

**Comment by pnuu** [comment]
In https://github.com/pytroll/pytroll-collectors/pull/157/commits/962e875596646b436b322cc53d736a25f2d8b28c I mock the trigger creation. I think I need to create a more involved mock or fake trigger creation with collectors to simulate the structure so that the checks in `_keep_running()` are there.

**Comment by pnuu** [comment]
With https://github.com/pytroll/pytroll-collectors/pull/157/commits/e866900bce5e5075a7d616eaefada426bc9f742a I think the new feature is "fully" tested.

**Review by mraspaud — APPROVED** [review]
LGTM!

**Pull request merged**